### PR TITLE
EVAKA-4142 citizen warnings when attendance differs from reservation

### PR DESCRIPTION
**Pull request by jaahon**

### CODE REVIEW

**Review by kimifoud — APPROVED** [review]

File: frontend/src/citizen-frontend/calendar/DayView.tsx [hunk 1]
```diff
@@ -617,3 +646,8 @@ const AbsenceButton = styled(InlineButton)`
     padding: ${defaultMargins.s};
   }
 `
+
+const Warning = styled.div`
+  grid-column: 1 / 3;
+  color: ${({ theme }) => theme.colors.accents.a2orangeDark};
```
[the remark this answers is not in the document]
Comment (jaahon) [reply frontend/src/citizen-frontend/calendar/DayView.tsx:652]: I guess the problem is that `colors.status.warning` doesn't have enough contrast to be used in text, but then the darker orange does not look good with icons. Maybe we need separate `icon` and `text` colors?

File: frontend/src/lib-common/reservations.ts [hunk 2]
```diff
@@ -195,3 +195,50 @@ export function validateTimeRange(timeRange: TimeRange): TimeRangeErrors {
...
+  attendances: TimeRange[]
+): boolean {
+  return reservations.some((reservation) => {
+    const matchingAttendance = attendances.find(
+      (attendance) =>
+        attendance.startTime <= reservation.endTime &&
+        reservation.startTime <= attendance.endTime
+    )
+
+    if (matchingAttendance == undefined) return false
```
[the remark this answers is not in the document]
Comment (jaahon) [reply frontend/src/lib-common/reservations.ts:234]: Yeah definitely not `==` 😅

File: service/src/main/kotlin/fi/espoo/evaka/reservations/ReservationControllerCitizen.kt [hunk 3]
```diff
@@ -199,6 +195,17 @@ LEFT JOIN LATERAL (
...
+    SELECT
+        jsonb_agg(
+            jsonb_build_object(
+                'startTime', to_char(greatest(ca.arrived AT TIME ZONE 'Europe/Helsinki', t), 'HH24:MI'),
+                'endTime', to_char(least(ca.departed AT TIME ZONE 'Europe/Helsinki', t + interval '1 day' - interval '1 minute'), 'HH24:MI')
+            ) ORDER BY ca.arrived ASC
+        ) AS attendances
+    FROM child_attendance ca WHERE ca.child_id = g.child_id
+    AND daterange((ca.arrived AT TIME ZONE 'Europe/Helsinki')::date, (ca.departed AT TIME ZONE 'Europe/Helsinki')::date, '[]') @> t::date
+) ca ON true
```
[the remark this answers is not in the document]
Comment (jaahon) [reply service/src/main/kotlin/fi/espoo/evaka/reservations/ReservationControllerCitizen.kt:208]: Yeah. If it's way too slow it should probably be moved to a separate request.

File: frontend/src/citizen-frontend/calendar/DayView.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (kimifoud) [reply frontend/src/citizen-frontend/calendar/DayView.tsx:652]: Yeah, that's true. I guess the UX department can decide how to proceed in the long run. Meanwhile, we could export a warningTextColor or something which could be warning color with x% of shade applied to it?

**Pull request merged**